src/main/scala/com/microsoft/hyperspace/index/dataskipping/DataSkippingIndex.scala

```diff
@@ -0,0 +1,132 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.index.dataskipping
+
+import org.apache.spark.sql._
+import org.apache.spark.sql.functions.input_file_name
+
+import com.microsoft.hyperspace.index._
+import com.microsoft.hyperspace.index.dataskipping.sketch.Sketch
+
+case class DataSkippingIndex(
+    sketches: Seq[Sketch],
+    sketchOffsets: Seq[Int],
+    override val properties: Map[String, String] = Map.empty)
+    extends Index {
+  assert(sketches.nonEmpty)
+  assert(sketches.length + 1 == sketchOffsets.length)
+
+  override def kind: String = DataSkippingIndex.kind
+
+  override def kindAbbr: String = DataSkippingIndex.kindAbbr
+
+  override def indexedColumns: Seq[String] = sketches.flatMap(_.columns)
+
+  override def referencedColumns: Seq[String] = indexedColumns
+
+  override def withNewProperties(newProperties: Map[String, String]): DataSkippingIndex = {
+    copy(properties = newProperties)
+  }
+
+  override def statistics(extended: Boolean = false): Map[String, String] = {
+    Map("sketches" -> sketches.mkString(", "))
+  }
+
+  override def canHandleDeletedFiles: Boolean = true
+
+  override def write(ctx: IndexerContext, indexData: DataFrame): Unit = {
+    indexData.write.mode(SaveMode.Overwrite).parquet(ctx.indexDataPath.toString)
+  }
+
+  override def optimize(ctx: IndexerContext, indexDataFilesToOptimize: Seq[FileInfo]): Unit = {
+    val indexData = ctx.spark.read.parquet(indexDataFilesToOptimize.map(_.name): _*)
+    write(ctx, indexData.repartition(1))
+  }
+
+  override def refreshIncremental(
+      ctx: IndexerContext,
+      appendedSourceData: Option[DataFrame],
+      deletedSourceDataFiles: Seq[FileInfo],
+      indexContent: Content): (Index, Index.UpdateMode) = {
+    if (appendedSourceData.nonEmpty) {
+      write(ctx, index(ctx, appendedSourceData.get))
+    }
+    (this, Index.UpdateMode.Merge)
+  }
+
+  override def refreshFull(
+      ctx: IndexerContext,
+      sourceData: DataFrame): (DataSkippingIndex, DataFrame) = {
+    val updatedIndex = copy(sketches = DataSkippingIndex.resolveSketch(sketches, sourceData))
+    (updatedIndex, updatedIndex.index(ctx, sourceData))
+  }
+
+  /**
+   * Creates index data for the given source data.
+   */
+  def index(ctx: IndexerContext, sourceData: DataFrame): DataFrame = {
+    val aggregateFunctions = sketches.flatMap { s =>
+      val aggrs = s.aggregateFunctions
+      assert(aggrs.nonEmpty && aggrs.length == s.numValues)
+      aggrs.map(aggr => aggr.as(DataSkippingIndex.normalizeColumnName(aggr.toString)))
+    }
+    val fileNameCol = "input_file_name"
+    val indexData = sourceData
+      .groupBy(input_file_name().as(fileNameCol))
```

Comment:
Could you add an example of index data to the PR description? e.g.
```
val df = spark.read.parquet("/path/to/index/v__=0")
df.limit(10).show(false)
```

Reply:
Done